jiant/proj/main/modeling/heads.py

```diff
@@ -4,14 +4,84 @@
 import torch.nn as nn
 
 import transformers
+
 from jiant.ext.allennlp import SelfAttentiveSpanExtractor
+from jiant.shared.model_resolution import ModelArchitectures
+from jiant.tasks.core import TaskTypes
+
 
 """
 In HuggingFace/others, these heads differ slightly across different encoder models.
 We're going to abstract away from that and just choose one implementation.
 """
 
 
+class JiantHeadFactory:
```

Comment:
New task heads added here

Reply:
Thinking aloud:
- If we're going to make this a class method, we might as well add a registry as well (to support potentially user-created heads that can be dynamically added)
- Would it be better to combine the head+taskmodels, rather than have separate factories for head and task model?
- The EmbeddingModel logic can be split off into its own factory, potentially. Fine leaving it for now.

Reply:
- Great idea to have dynamically supported user heads. I'll look into the details of this (will set up a meeting for feedback).
- I was debating this as well. I am not set on this implementation but I chose this since it seemed better to isolate the components to prevent unwanted coupling (although they are closely linked already)
- Thanks

Reply:
My sense is that as long as the heads have their own standlone implementations, that's sufficiently isolated and tying the whole task-model construction should be fine.